docs/zh/parameter_configuration.md
# 参数配置<a name="ZH-CN_TOPIC_0000002496570833"></a>

## 环境变量列表<a name="ZH-CN_TOPIC_0000002496571433"></a>

**表 1**  MIS环境变量

|名称|类型|描述|取值范围|
|--|--|--|--|
|MIS_CACHE_PATH|str|缓存权重路径，需具有正确的权限与所属。|默认值：$HOME/mis/.cache。|
|MIS_MODEL|str|服务模型名称。|默认值：Qwen3-8B。<br>取值范围：[Qwen3-8B]。|
|MIS_ENGINE_TYPE|str|模型引擎类型。|默认值：vllm。<br>取值范围：[vllm]。|
|MIS_CONFIG|str|优化配置名称。|默认值：atlas800ia2-1x32gb-bf16-vllm-default。<br>取值范围请参考[模型支持与配置列表](#模型最优配置)。|
|MIS_PORT|int|服务绑定的端口。|默认值：8000。<br>取值范围：[1024, 65535]。|
|MIS_ENABLE_DOS_PROTECTION|bool|使能或去使能MIS的防DOS攻击特性。包含限制请求头/体大小、限制并发、限流、限制超时。|默认值：True。<br>当取值为“true”（忽略大小写）或“1”时设为True；其他值设为False。|
|MIS_LOG_LEVEL|str|MIS的日志等级。|默认值：INFO。<br>取值范围：[DEBUG, INFO, WARNING, ERROR, CRITICAL]。|
|MIS_MAX_LOG_LEN|int|配置日志的最大长度。|默认值：2048。<br>取值范围：[0, 8192]。|
|UVICORN_LOG_LEVEL|str|配置Uvicorn服务的日志级别。|默认值：info。<br>取值范围：[debug, info, warning, error, critical]。|

> [!NOTE] 说明
>三方组件的环境变量可能会更改MIS的服务状态，请参阅对应三方组件文档。

## 模型最优配置<a name="ZH-CN_TOPIC_0000002463252414"></a>

> [!NOTE] 说明
>
>- 请确保参数配置路径及文件与运行用户属主一致。
>- 请确保参数配置文件的大小不超过1MB。
>- 请确保参数配置文件不为软链接，路径字符串长度不大于1024。
>- 请确保参数配置路径权限为750，配置文件为640。

**模型支持<a name="section7482154712357"></a>**

**表 1**  模型支持与配置列表

|配置模型名|可选值|计算服务器硬件型号|计算卡规格|数据类型|后端|性能倾向|最低内存需求|
|--|--|--|--|--|--|--|--|
|qwen3-8b|atlas800ia2-1x32gb-bf16-vllm-default|Atlas 800I A2|1x32GB|BF16|vLLM|均衡|16GB|
|qwen3-8b|atlas800ia2-1x32gb-bf16-vllm-latency|Atlas 800I A2|1x32GB|BF16|vLLM|低时延|16GB|
|qwen3-8b|atlas800ia2-1x32gb-bf16-vllm-throughput|Atlas 800I A2|1x32GB|BF16|vLLM|高吞吐|16GB|

**优化配置参数示例<a name="section49261519367"></a>**

```shell
vllm:  # 引擎类型 (必填项，可选项为vllm，与后续engine_type对应)  
  dtype: bfloat16  # 具体配置信息(选填项，详情见如下配置参数总览)  
  tensor_parallel_size: 1  
  pipeline_parallel_size: 1  
  max_num_seqs: 128  
  block_size: 128
engine_type: vllm  # 必填项，可选项为vllm, 与引擎类型相应
model: Qwen3-8B  # 模型权重路径(必填项)
```

**优化配置参数列表<a name="section03608318374"></a>**

**表 2**  模型与引擎配置

|参数名|类型|描述|
|--|--|--|
|vllm|dict|引擎类型对应的配置字典。键名及其值请参考[表3 引擎详细配置](#table3)。|
|engine_type|str|引擎类型，取值范围：[vllm]。|
|model|str|模型的权重路径，取值范围：[Qwen3-8B]。|

**表 3**  引擎详细配置<a id="table3"></a>

|参数名|**类型**|**描述**|
|--|--|--|
|dtype|str|数据类型，取值范围：[bfloat16]。|
|tensor_parallel_size|int|张量并行大小，取值范围：[1, 2, 4, 8]。|
|pipeline_parallel_size|int|管道并行大小，取值范围：1。|
|distributed_exector_size|str|分布式执行器后端，取值范围：[mp]。|
|max_num_seqs|int|最大序列数，取值范围：[1, 512]。|
|max_model_len|int|最大模型长度（根据模型参数实际配置），取值范围：[1, 64000]。|
|max_num_batched_tokens|int|最大批处理词元数（根据模型参数实际配置），取值范围：[1, 64000]。|
|gpu_memory_utilization|float|GPU内存利用率，取值范围：[0.0, 1.0]。|
|block_size|int|块大小，取值范围：[16, 32, 64, 128]。|
|swap_space|int|交换空间大小（GB），取值范围：[0, 1024]。|
|cpu_offload_gb|int|CPU卸载大小（GB），取值范围：[0, 1024]。|
|scheduling_policy|str|调度策略，取值范围：[fcfs, priority]。|
|num_scheduler_steps|int|多步调度器数，取值范围：[1, 1024]。|
|enable_chunked_prefill|bool|是否启用分块预填充，取值范围：[True, False]。|
|enable_prefix_caching|bool|是否启用前缀缓存，取值范围：[True, False]。|
|multi_step_stream_outputs|bool|是否启用多步流输出，取值范围：[True, False]。|
|enforce_eager|bool|是否强制即时执行模式，取值范围：[True, False]。|


mis/llm/engines/config.py
# -*- coding:utf-8 -*-
# Copyright (c) Huawei Technologies Co. Ltd. 2025. All rights reserved.
import os
import sys
from abc import ABC
from typing import Callable, Dict, Optional, Type, Union

import yaml

from mis.args import GlobalArgs
from mis.constants import HW_910B, MIS_ENGINE_TYPES, MIS_MODEL_LIST, MIS_MAX_CONFIG_SIZE
from mis.logger import init_logger
from mis.utils.utils import ConfigChecker, get_soc_name

logger = init_logger(__name__)

MIS_CONFIG_DEFAULT = {
    "qwen3-8b": {HW_910B: "atlas800ia2-1x32gb-bf16-vllm-default"},
}

CHECKER_VLLM = {
    "dtype": {
        "type": "str_in",
        "valid_values": ("bfloat16",)
    },
    "tensor_parallel_size": {
        "type": "int",
        "valid_values": (1, 2, 4, 8)
    },
    "pipeline_parallel_size": {
        "type": "int",
        "valid_values": (1,)
    },
    "distributed_executor_backend": {
        "type": "str_in",
        "valid_values": ("mp",)
    },
    "max_num_seqs": {
        "type": "int",
        "min": 1,
        "max": 512
    },
    "max_model_len": {
        "type": "int",
        "min": 1,
        "max": 64000
    },
    "max_num_batched_tokens": {
        "type": "int",
        "min": 1,
        "max": 64000
    },
    "max_seq_len_to_capture": {
        "type": "int",
        "min": 1,
        "max": 64000
    },
    "gpu_memory_utilization": {
        "type": "float",
        "min": 0.0,
        "max": 1.0
    },
    "block_size": {
        "type": "int",
        "valid_values": (16, 32, 64, 128)
    },
    "swap_space": {
        "type": "int",
        "min": 0,
        "max": 1024,
    },
    "cpu_offload_gb": {
        "type": "int",
        "min": 0,
        "max": 1024
    },
    "scheduling_policy": {
        "type": "str_in",
        "valid_values": ("fcfs", "priority")
    },
    "num_scheduler_steps": {
        "type": "int",
        "min": 1,
        "max": 1024
    },
    "enable_chunked_prefill": {
        "type": "bool",
        "valid_values": (True, False)
    },
    "enable_prefix_caching": {
        "type": "bool",
        "valid_values": (True, False)
    },
    "multi_step_stream_outputs": {
        "type": "bool",
        "valid_values": (True, False)
    },
    "enforce_eager": {
        "type": "bool",
        "valid_values": (True, False)
    },
}


class AbsEngineConfigValidator(ABC):
    _engine_config_validation: Dict[str, Type["AbsEngineConfigValidator"]] = {}

    def __init__(self, config: Dict, checkers: Dict) -> None:
        """
        Initialize the AbsEngineConfigValidator class.
        :param config: The configuration dictionary.
        :param checkers: The checker dictionary OF selected engine.
        """
        self.config = config
        self.checkers = checkers
        diff_config = set(self.config.keys()) - set(self.checkers.keys())
        if diff_config:
            # Check if the backend configuration keywords for inference are updated with the version
            logger.debug(f"Configuration keys {diff_config} are not supported. ")
        self.config_update: Dict = {key: self.config[key] for key in self.config if key in self.checkers.keys()}

    @classmethod
    def register(cls, engine_type: str) -> Callable:
        """
        Register an engine configuration validator.
        :param engine_type: Engine type.
        """

        def decorator(subclass):
            cls._engine_config_validation[engine_type] = subclass
            return subclass

        return decorator

    @classmethod
    def get_validator(cls, engine_type: str) -> Type["AbsEngineConfigValidator"]:
        """
        Get the engine configuration validator.
        :param engine_type: Engine type.
        :return: Engine configuration validator class.
        """
        return cls._engine_config_validation.get(engine_type)

    def filter_and_validate_config(self) -> Dict:
        """
        Filter and validate the configuration.
        :return: Right configuration.
        """
        logger.debug("Filtering and validating configuration.")
        valid_config = {}
        for key, value in self.config_update.items():
            checker = self.checkers.get(key)

            is_valid = True
            if "valid_values" in checker:
                is_valid = ConfigChecker.is_value_in_enum(key, value, checker.get("valid_values"))
            elif "min" in checker and "max" in checker:
                is_valid = ConfigChecker.is_value_in_range(key, value, checker.get("min"), checker.get("max"))

            if is_valid:
                logger.debug(f"Configuration item {key} is valid.")
                valid_config[key] = value
        logger.debug("Configuration filtered and validated successfully.")
        return valid_config


@AbsEngineConfigValidator.register("vllm")
class VLLMEngineConfigValidator(AbsEngineConfigValidator):
    """
    VLLM engine configuration validator.
    """

    def __init__(self, config: Dict) -> None:
        """
        vLLM Engine configuration validator initialization.
        :param config: Configuration parameters.
        """
        super().__init__(config, CHECKER_VLLM)


class ConfigParser:
    def __init__(self, args: GlobalArgs) -> None:
        """
        Check all args are valid.
        :param args: global args
        """
        logger.debug(f"Initializing ConfigParser with args.")
        self.args = args
        self.args.engine_optimization_config = {}

        self._check_all_args_valid()

        self.model_type = self.args.model.split('/')[-1]
        self.model_folder_path = os.path.join(self._get_configs_root(), "configs", "llm", self.model_type.lower())
        self.engine_type = self.args.engine_type
        self.mis_config = self.args.mis_config
        logger.debug(f"ConfigParser initialized successfully")

    @staticmethod
    def _get_configs_root():
        """Get the root path of configs, which may be at the same level as the startup script or one level above it."""
        try:
            if not sys.argv[0]:
                logger.error("Failed to get the script path.")
                raise Exception("Failed to get the script path.")
            script_dir = os.path.dirname(os.path.realpath(sys.argv[0]))
            parent_dir = os.path.dirname(script_dir)
            configs_dir = os.path.join(script_dir, "configs")
            if not os.path.isdir(configs_dir):
                script_dir = parent_dir
        except OSError as e:
            logger.error(f"Failed to get the configs root directory: {e}")
            raise OSError("Failed to get the configs root directory.") from e
        if script_dir is None:
            logger.error("Failed to get the configs root directory.")
            raise Exception("Failed to get the configs root directory.")
        return script_dir

    @staticmethod
    def _config_attr_update(selected_engine_type: str, selected_engine_config: Dict) -> Dict:
        """
        Update the attributes of the selected engine.
        :param selected_engine_type: The type of the selected engine.
        :param selected_engine_config: The configuration of the selected engine.
        :return: updated config dictionary.
        """
        logger.debug(f"Updating attributes for engine type: {selected_engine_type}.")
        validator_class = AbsEngineConfigValidator.get_validator(selected_engine_type)
        validator = validator_class(selected_engine_config.get(selected_engine_type))
        logger.debug(f"Attributes for engine type {selected_engine_type} updated successfully.")
        return validator.filter_and_validate_config()

    def engine_config_loading(self) -> GlobalArgs:
        """
        Obtain the engine configuration. If the parameters are successfully obtained, update the args.
        :return: Update global parameters
        """
        logger.debug("Loading engine configuration.")
        engine_optimization_config = self._load_config_from_file()
        if not engine_optimization_config or not self._is_config_valid(engine_optimization_config):
            logger.warning("Using default engine configuration.")
            return self.args

        engine_type_selected = engine_optimization_config.get("engine_type")
        self.args.engine_optimization_config = self._config_attr_update(engine_type_selected,
                                                                        engine_optimization_config)

        model_by_config = engine_optimization_config.get("model")
        self.args.model = model_by_config if model_by_config in MIS_MODEL_LIST else self.args.model
        self.args.engine_type = engine_type_selected
        logger.info("Engine configuration loaded successfully.")
        return self.args

    def _config_yaml_file_loading(self, config_file_path: str) -> Union[Dict, None]:
        """
        Load config file.
        :params config_file_path: config path
        :return: Parsed configuration dictionary if successful, None otherwise.
        """
        logger.debug(f"Loading configuration: {self.mis_config}.")
        try:
            if not os.path.isfile(config_file_path):
                logger.error(f"The configuration {self.mis_config} does not exist.")
                raise Exception("The configuration file does not exist.")
            if os.path.islink(config_file_path):
                logger.error(f"The configuration {self.mis_config} is a symbolic link.")
                raise Exception("The configuration file is a symbolic link.")

            file_size = os.path.getsize(config_file_path)
            if file_size > MIS_MAX_CONFIG_SIZE:
                logger.error(f"The size of the configuration {self.mis_config} exceeds 1MB.")
                raise Exception("The size of the configuration file exceeds 1MB.")

            with open(config_file_path, "r", encoding="utf-8") as file:
                config = yaml.safe_load(file)
                logger.info(f"Configuration {self.mis_config} loaded successfully.")

            return config

        except FileNotFoundError:
            logger.warning(f"Config {self.mis_config} not found. "
                         "The engine will be started with default parameters.")
            return None
        except (OSError, yaml.YAMLError) as e:
            logger.error(f"Failed to load configuration {self.mis_config}: {e}")
            raise e
        except Exception as e:
            logger.error(f"Unexpected error while loading configuration {self.mis_config}: {e}")
            raise e

    def _load_config_from_file(self) -> Optional[Dict]:
        """
        Loading the configuration from a specified YAML file.
        If the specified configuration file does not exist or the environment variable `MIS_CONFIG` is not set,
        the method will use the default configuration.

        :returns: A dictionary containing the configuration loaded from the YAML file,
        or `None` if no valid configuration is found.
        """
        logger.debug("Loading configuration from file.")
        if self.mis_config is None:
            logger.warning("The environment variable MIS_CONFIG is missed. "
                           "Please check if the environment variables is valid."
                           "The engine will be started with the default parameters.")
            self.mis_config = self._get_default_config()

        elif not os.path.exists(os.path.join(self.model_folder_path, self.mis_config + ".yaml")):
            logger.debug(f"Selected config {self.mis_config} does not exist."
                         f"The engine will be started with the default config.")
            self.mis_config = self._get_default_config()

        if self.mis_config is None:
            logger.warning("No valid configuration found. Using default parameters.")
            return None

        config_file_path = os.path.join(self.model_folder_path, self.mis_config + ".yaml")
        engine_optimization_config = self._config_yaml_file_loading(config_file_path)
        logger.debug("Configuration file loaded successfully.")
        return engine_optimization_config

    def _is_config_valid(self, config: Dict) -> bool:
        """
        Checks if the config is valid.
        :param config: Config to check.
        :return: True if the config is valid, False otherwise.
        """
        logger.debug("Validating configuration.")
        if config is None or not isinstance(config, dict):
            logger.debug(f"The YAML config file for {self.model_type} ({self.mis_config}) is invalid."
                         f"The engine will be started with the default parameters.")
            return False

        if config.get("engine_type") is None or config.get("model") is None:
            logger.debug(f"Please check if keywords engine_type and model in "
                         f"{self.model_type} ({self.mis_config}) YAML is complete."
                         f"The engine will be started with the default parameters.")
            return False

        engine_type = config.get("engine_type")
        if engine_type not in MIS_ENGINE_TYPES:
            logger.warning(f"engine_type in YAML config file must in {MIS_ENGINE_TYPES}")
            return False

        if config.get(engine_type) is None:
            logger.warning(f"Config of engine type {config.get('engine_type')} is required. "
                         f"The engine will be started with the default parameters.")
            return False
        logger.debug("Configuration is valid.")
        return True

    def _check_all_args_valid(self) -> None:
        """
        Check all args are valid. If not, raise ValueError
        """
        logger.debug("Checking all arguments.")
        if not isinstance(self.args, GlobalArgs):
            logger.error("args must be an instance of GlobalArgs")
            raise TypeError("args must be an instance of GlobalArgs")

        for attr in ("engine_type", "mis_config"):
            # Verify the attribute character string.
            args_attr = getattr(self.args, attr, None)
            if args_attr is not None:
                try:
                    ConfigChecker.check_string_input(attr, args_attr)
                except Exception as e:
                    raise Exception(f"{attr} in args is not a valid string: {e}") from e
        logger.info("All arguments are valid.")

    def _get_default_config(self) -> Union[str, None]:
        logger.info("Getting default configuration.")
        if self.model_type.lower() not in MIS_CONFIG_DEFAULT:
            logger.info(f"The current model {self.model_type} is not included in the optimization configuration "
                        f"or is not a large model (e.g., a embedding model).")
            return None

        soc_name = get_soc_name()
        model_config_dict = MIS_CONFIG_DEFAULT.get(self.model_type.lower())
        if model_config_dict is None:
            logger.warning(f"Default configuration for {self.model_type} is not found. "
                           f"Please check whether the model name is consistent with the image description.")
            return None
        model_config_default = model_config_dict.get(soc_name)
        if model_config_default is None:
            logger.warning(f"The current model {self.model_type} is not compatible with "
                           f"the hardware platform {soc_name}. "
                           f"Check the hardware support status of the model in the image description.")
            return None
        logger.info(f"Default configuration for {self.model_type} on {soc_name} is {model_config_default}.")
        return model_config_default
